File: backend/app/routers/cases.py
import random
from datetime import datetime
from typing import Optional
from fastapi import APIRouter, HTTPException, Query, status
from app.services.storage import db
from app.schemas import CaseCreate, CaseUpdate, CaseResponse, CasesListResponse

router = APIRouter(prefix="/cases", tags=["Cases"])

@router.get("", response_model=CasesListResponse)
def get_cases(
    search: Optional[str] = Query(None),
    status: Optional[str] = Query(None),
    riskLevel: Optional[str] = Query(None)
):
    filtered = list(db.cases)

    if search:
        q = search.lower()
        filtered = [
            c for c in filtered
            if q in c["id"].lower()
            or q in c["title"].lower()
            or q in c["client"].lower()
            or q in c["type"].lower()
        ]

    if status and status != "All":
        filtered = [c for c in filtered if c["status"] == status]

    if riskLevel and riskLevel != "All":
        filtered = [c for c in filtered if c["riskLevel"] == riskLevel]

    return {
        "success": True,
        "count": len(filtered),
        "data": filtered
    }

@router.get("/{case_id}")
def get_case_by_id(case_id: str):
    case_item = next((c for c in db.cases if c["id"] == case_id), None)
    if not case_item:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Case with ID {case_id} not found"
        )
    return {"success": True, "data": case_item}

@router.post("", status_code=status.HTTP_201_CREATED)
def create_case(payload: CaseCreate):
    new_case = {
        "id": f"CASE-2026-{random.randint(1000, 9999)}",
        "title": payload.title,
        "client": payload.client,
        "court": payload.court or "Standard Jurisdiction",
        "type": payload.type or "General Legal Trace",
        "status": payload.status or "In Review",
        "riskLevel": payload.riskLevel or "Medium",
        "complianceScore": payload.complianceScore or random.randint(70, 95),
        "leadCounsel": payload.leadCounsel or "Unassigned",
        "lastUpdated": datetime.now().isoformat(),
        "auditCount": 1,
        "summary": payload.summary or "Newly registered legal trace file."
    }
    db.cases.insert(0, new_case)
    return {"success": True, "data": new_case}

@router.patch("/{case_id}")
def update_case(case_id: str, payload: CaseUpdate):
    case_idx = next((i for i, c in enumerate(db.cases) if c["id"] == case_id), -1)
    if case_idx == -1:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Case with ID {case_id} not found"
        )
    
    update_data = payload.model_dump(exclude_unset=True)
    update_data["lastUpdated"] = datetime.now().isoformat()
    db.cases[case_idx].update(update_data)
    
    return {"success": True, "data": db.cases[case_idx]}

@router.delete("/{case_id}")
def delete_case(case_id: str):
    case_idx = next((i for i, c in enumerate(db.cases) if c["id"] == case_id), -1)
    if case_idx == -1:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Case with ID {case_id} not found"
        )
    
    deleted = db.cases.pop(case_idx)
    return {
        "success": True,
        "data": deleted,
        "message": f"Case {case_id} successfully archived/removed"
    }

File: backend/routes/cases.js
import express from 'express';

const router = express.Router();

// Mock database store for legal cases
let cases = [
  {
    id: 'CASE-2026-0891',
    title: 'Apex Financial vs. Meridian Corp',
    client: 'Apex Financial Services',
    court: 'Delaware Chancery Court / High Court',
    type: 'Corporate Governance & Due Diligence',
    status: 'In Review',
    riskLevel: 'High',
    complianceScore: 84,
    leadCounsel: 'Elena Rostova, Esq.',
    lastUpdated: '2026-08-28T14:30:00Z',
    auditCount: 14,
    summary: 'Cross-border entity structure audit, contract ownership verification, and regulatory disclosure tracing.'
  },
  {
    id: 'CASE-2026-0742',
    title: 'EU AI Act Governance Certification',
    client: 'NeuralTech Dynamics',
    court: 'European AI Safety Board',
    type: 'AI Governance & Compliance',
    status: 'Compliant',
    riskLevel: 'Low',
    complianceScore: 97,
    leadCounsel: 'Marcus Vance, Esq.',
    lastUpdated: '2026-08-29T09:15:00Z',
    auditCount: 32,
    summary: 'Algorithmic impact assessment, training data lineage audit, and human-in-the-loop trace record.'
  },
  {
    id: 'CASE-2026-0618',
    title: 'Vanguard Holdings Acquisition Trace',
    client: 'Vanguard Capital',
    court: 'FTC Antitrust Division',
    type: 'M&A Due Diligence',
    status: 'Pending Evidence',
    riskLevel: 'Medium',
    complianceScore: 78,
    leadCounsel: 'Sarah Jenkins, Esq.',
    lastUpdated: '2026-08-27T18:45:00Z',
    auditCount: 21,
    summary: 'Data room structure analysis, key IP ownership verification, and regulatory reporting.'
  },
  {
    id: 'CASE-2026-0455',
    title: 'BioHealth IP Rights Litigation',
    client: 'BioHealth Innovations',
    court: 'U.S. Federal District Court',
    type: 'IP & Patent Rights',
    status: 'Active Discovery',
    riskLevel: 'High',
    complianceScore: 69,
    leadCounsel: 'David K. Chen',
    lastUpdated: '2026-08-29T11:00:00Z',
    auditCount: 18,
    summary: 'Patent claim trace, prior art search validation, and expert witness testimony tracking.'
  }
];

// GET /api/cases
router.get('/', (req, res) => {
  const { search, status, riskLevel } = req.query;
  let filtered = [...cases];

  if (search) {
    const q = search.toLowerCase();
    filtered = filtered.filter(c => 
      c.id.toLowerCase().includes(q) ||
      c.title.toLowerCase().includes(q) ||
      c.client.toLowerCase().includes(q) ||
      c.type.toLowerCase().includes(q)
    );
  }

  if (status && status !== 'All') {
    filtered = filtered.filter(c => c.status === status);
  }

  if (riskLevel && riskLevel !== 'All') {
    filtered = filtered.filter(c => c.riskLevel === riskLevel);
  }

  res.json({
    success: true,
    count: filtered.length,
    data: filtered
  });
});

// GET /api/cases/:id
router.get('/:id', (req, res) => {
  const item = cases.find(c => c.id === req.params.id);
  if (!item) {
    return res.status(404).json({ success: false, message: 'Case not found' });
  }
  res.json({ success: true, data: item });
});

// POST /api/cases
router.post('/', (req, res) => {
  const { title, client, court, type, riskLevel, leadCounsel, summary } = req.body;
  if (!title || !client) {
    return res.status(400).json({ success: false, message: 'Title and Client are required fields' });
  }

  const newCase = {
    id: `CASE-2026-${Math.floor(1000 + Math.random() * 9000)}`,
    title,
    client,
    court: court || 'Standard Jurisdiction',
    type: type || 'General Legal Trace',
    status: 'In Review',
    riskLevel: riskLevel || 'Medium',
    complianceScore: Math.floor(70 + Math.random() * 25),
    leadCounsel: leadCounsel || 'Unassigned',
    lastUpdated: new Date().toISOString(),
    auditCount: 1,
    summary: summary || 'Newly registered legal trace file.'
  };

  cases.unshift(newCase);
  res.status(201).json({ success: true, data: newCase });
});

// PATCH /api/cases/:id
router.patch('/:id', (req, res) => {
  const index = cases.findIndex(c => c.id === req.params.id);
  if (index === -1) {
    return res.status(404).json({ success: false, message: 'Case not found' });
  }

  cases[index] = {
    ...cases[index],
    ...req.body,
    lastUpdated: new Date().toISOString()
  };

  res.json({ success: true, data: cases[index] });
});

// DELETE /api/cases/:id
router.delete('/:id', (req, res) => {
  const index = cases.findIndex(c => c.id === req.params.id);
  if (index === -1) {
    return res.status(404).json({ success: false, message: 'Case not found' });
  }

  const deleted = cases.splice(index, 1)[0];
  res.json({ success: true, data: deleted, message: 'Case successfully archived/removed' });
});

export default router;
